[4][Span logging] Add conversion utils to / from MLflow Spans <> OTel Spans

dbczumar opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/dbczumar/mlflow/pull/17115?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17115/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17115/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/17115/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Add conversion utils to / from MLflow Spans <> OTel Spans

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

dbczumar committed:
Add REST store and API support for logging spans

- Add log_spans abstract method to AbstractStore
- Implement log_spans in RestStore using OpenTelemetry protocol
- Add Span entity from insights branch
- Add minimal FastAPI app wrapper and OTel API endpoint handler

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Revert span.py to origin/master and remove trace_state handling

- Revert mlflow/entities/span.py to origin/master version
- Remove trace_state handling from rest_store.py as it's unnecessary

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Refactor log_spans to use cleaner conversion method

- Extract MLflow to OTel span conversion into separate _convert_mlflow_span_to_otel_span method
- Simplify log_spans method by using the conversion helper
- Improve code organization and readability

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Improve _convert_mlflow_span_to_otel_span to return OTel span

- Method now creates and returns OTel span instead of requiring one as input
- Cleaner API with more intuitive usage
- log_spans now uses the returned span directly

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move OTel proto conversion to Span entity

- Add _to_otel_proto() method to Span class for OpenTelemetry protobuf conversion
- Remove _convert_mlflow_span_to_otel_span from rest_store.py
- Cleaner separation of concerns with conversion logic on the entity itself
- Remove unused SpanStatusCode import from rest_store.py

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add _to_otel_proto method to SpanEvent

- Add _to_otel_proto() method to SpanEvent for OpenTelemetry protobuf conversion
- Update Span._to_otel_proto() to use SpanEvent._to_otel_proto()
- Better separation of concerns with each entity handling its own conversion

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove _convert_otel_span_to_mlflow_span from otel_api.py

- Remove the complex conversion function as requested
- Simplify otel_api.py to be a minimal placeholder
- Add TODO comment for future implementation of OTel JSON to MLflow span conversion
- Remove unused imports

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use existing helper functions for ID encoding in _to_otel_proto

- Replace inline to_bytes() calls with existing _encode_trace_id_to_byte() and _encode_span_id_to_byte() helpers
- Avoid code duplication and use consistent encoding logic

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix unnecessary ID conversions in _to_otel_proto

- Use internal _span.context.trace_id and span_id directly (already integers)
- Avoid converting hex string to int just to convert back to bytes
- Use _span.parent.span_id directly for parent ID
- Cleaner and more efficient code

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify ID conversion in _to_otel_proto

- Use bytes.fromhex() to convert hex string IDs directly to bytes
- Remove unnecessary use of helper functions
- Use self._trace_id, self.span_id, and self.parent_id directly
- Much simpler and cleaner code

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add _from_otel_proto class method to Span

- Add method to create Span from OpenTelemetry protobuf format
- Converts OTel proto span to OTelReadableSpan then wraps in Span
- Useful for receiving spans via OTel protocol in the future

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add _bytes_to_id helper function for protobuf conversion

- Add _bytes_to_id helper to convert raw bytes to integer IDs
- Use the helper in _from_otel_proto for cleaner code
- Consistent approach for bytes to ID conversion

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move imports to top of rest_store.py

- Move google.protobuf and opentelemetry imports to top of file
- Add http_request and verify_rest_response to rest_utils imports
- Remove duplicate imports from inside log_spans method

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add REST store and REST API changes for logging spans

- Add log_spans and log_spans_async abstract methods to AbstractStore
- Implement REST store log_spans methods using OpenTelemetry protocol
- Add opentelemetry-proto dependency to requirements files
- Convert MLflow spans to OTel protobuf format for REST API submission

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Improve REST store protobuf attribute handling

- Simplify protobuf syntax in rest_store.py by removing unnecessary CopyFrom
- Add proper type conversion helpers for OTel protobuf attributes
- Use WhichOneof pattern with getattr for cleaner decoding
- Support all OTel attribute types (string, int, float, bool, arrays, dicts)

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add OTel protobuf converters to SpanStatus and use in Span converters

- Add to_otel_proto_status() and from_otel_proto_status() methods to SpanStatus
- Use SpanStatus converters in Span._to_otel_proto() and _from_otel_proto()
- Remove redundant status conversion code from Span
- Clean up unused imports

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move imports to top of file in span.py

- Move opentelemetry.proto.trace.v1.trace_pb2 import to top
- Remove duplicate imports from within methods
- Use consistent import naming (OTelProtoSpan for protobuf Span)

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Merge remote-tracking branch 'origin/master' into span_otel_conversion

# Conflicts:
#	mlflow/server/fastapi_app.py
#	mlflow/store/tracking/abstract_store.py

dbczumar committed:
Add Span <> OTel protobuf conversion utilities

- Add _set_value_on_otel_proto helper for encoding Python values to OTel AnyValue
- Add _decode_otel_value helper for decoding OTel AnyValue to Python
- Add _to_otel_proto method to Span for OTLP export format
- Add _from_otel_proto classmethod to create Span from OTel protobuf
- Add to_otel_proto_status and from_otel_proto_status to SpanStatus
- Update SpanEvent._to_otel_proto to use new helpers
- Use WhichOneof pattern for cleaner protobuf field checking

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify event and attribute conversion using comprehensions

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove unnecessary if checks for status and events

Status is always defined (defaults to UNSET) and events is always a list (can be empty)

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove obvious comments from conversion methods

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Pass attributes and events directly to OTelReadableSpan constructor

Remove unnecessary intermediate variables

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add explicit AnyValue type hints for OTel protobuf conversion functions

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Convert OTel proto status directly to OTel SDK status

Remove unnecessary round-trip conversion through MLflow SpanStatus

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add return type hint to _to_otel_proto method

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Rename _bytes_to_id to _otel_proto_bytes_to_id for clarity

The function is specifically for converting OTel protobuf byte representations

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move OTel protobuf conversion utilities to mlflow/tracing/utils/otlp.py

- Move set_otel_proto_anyvalue and decode_otel_proto_anyvalue to otlp.py
- Move otel_proto_bytes_to_id to otlp.py
- Update all imports to use the centralized utilities
- Better code organization with OTel-related utilities in one place

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move imports to top of files and remove unnecessary checks

- Move all imports to top of files in span.py, span_event.py, span_status.py
- Remove unnecessary if check for attributes in span_event.py
- Clean up redundant imports inside methods

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use OTel proto Status constants instead of magic numbers

Replace magic numbers (1, 2) with readable constants like
Status.STATUS_CODE_OK and Status.STATUS_CODE_ERROR

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add comprehensive tests for Span <-> OTel protobuf conversion

- Added thorough test coverage for Span to/from OTel protobuf conversion
- Tests cover status, attributes, events, and roundtrip conversion
- Added opentelemetry-proto dependency to requirements
- Fixed status message handling for non-ERROR statuses in tests
- All tests passing with proper validation of conversion logic

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add opentelemetry-proto dependency to mlflow-tracing package

The tracing package also needs opentelemetry-proto since it uses the
OTel protobuf conversion utilities for span serialization.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Refactor test_span_otel_conversion to use functions instead of class

Per MLflow test conventions, tests should use functions rather than
classes. Refactored all test methods to be standalone test functions.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move OTel conversion tests to test_span.py

Per feedback, consolidate all span-related tests into the existing
test_span.py file rather than creating a separate test file. This
keeps all Span entity tests together in one place.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move OTel imports to top of test_span.py

Follow Python convention of having all imports at the top of the file
rather than scattered throughout. Remove duplicate imports that were
left in the middle of the file.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Reorganize OTel conversion tests with clearer names

- Move status conversion tests to test_span_status.py where they belong
- Move span event conversion test to test_span_event.py
- Rename tests with clearer, more descriptive names:
  - test_otel_status_to_mlflow_status_conversion (OTel SDK to MLflow)
  - test_otel_proto_status_to_mlflow_status_conversion (OTel proto to MLflow)
  - test_mlflow_status_to_otel_proto_status_conversion (MLflow to OTel proto)
  - test_span_event_to_otel_proto_conversion (SpanEvent to OTel proto)
- Keep only span-specific tests in test_span.py

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove unnecessary comment from test_span.py

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Enhance roundtrip test to verify span attributes are preserved

Added comprehensive verification that span attributes are properly
preserved during OTel protobuf roundtrip conversion, including:
- Custom attributes (like custom_attr)
- Trace request ID
- Span type, inputs, and outputs

This ensures the conversion properly handles all span attributes,
not just event attributes.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Verify ALL attributes in roundtrip conversion test

Enhanced test_otel_roundtrip_conversion to iterate through ALL span
attributes and verify each one is preserved correctly during the
OTel protobuf roundtrip conversion. This ensures comprehensive
coverage of attribute preservation.

The test now:
- Iterates through all attributes in the original span
- Verifies each attribute exists in the roundtrip span
- Verifies each attribute value is preserved exactly
- Still includes explicit checks for important attributes like
  custom_attr and trace request ID for clarity

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add :meta private: directive to internal OTel conversion methods

Added :meta private: directive to mark internal methods that should
not appear in public documentation:
- span_status.py: to_otel_proto_status, from_otel_proto_status
- otlp.py: otel_proto_bytes_to_id, set_otel_proto_anyvalue,
  decode_otel_proto_anyvalue

These methods are implementation details for OTel protobuf conversion
and should not be part of the public API.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add :meta private: to SDK status conversion methods

Added :meta private: directive to the OpenTelemetry SDK status
conversion methods in span_status.py:
- to_otel_status: Converts to OTel SDK Status
- from_otel_status: Converts from OTel SDK Status

These are internal implementation details for OTel SDK integration
and should not be part of the public API documentation.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

harupy commented on mlflow/entities/span_status.py:
```diff
@@ -4,6 +4,7 @@
 from enum import Enum
 
 from opentelemetry import trace as trace_api
+from opentelemetry.proto.trace.v1.trace_pb2 import Status
```
```suggestion
from opentelemetry.proto.trace.v1.trace_pb2 import Status as OtelStatus
```

can we alias this?

dbczumar committed:
Address review comments: Alias OtelStatus and make function private

- Alias Status import as OtelStatus in span_status.py to disambiguate
- Add underscore prefix to otel_proto_bytes_to_id function to make it private
- Update all references to use the new names

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove :meta private: from new utility functions in otlp.py

Signed-off-by: dbczumar <corey.zumar@databricks.com>

harupy reviewed (approved):
LGTM!

harupy commented on mlflow/tracing/utils/otlp.py:
```diff
@@ -60,3 +62,68 @@ def _get_otlp_protocol() -> str:
+    return int.from_bytes(id_bytes, byteorder="big", signed=False)
+
+
+def set_otel_proto_anyvalue(pb_any_value: AnyValue, value: Any) -> None:
```
```suggestion
def _set_otel_proto_anyvalue(pb_any_value: AnyValue, value: Any) -> None:
```

can we make this private?

1. `dev/check_function_signatures.py` thinks this is a public function and warns you when a new required parameter is added.
2. If this function is imported in a documented module, this would also be documented.

dbczumar replied:
Absolutely! Done!

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>
